File: llm.go
package main

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"regexp"
	"strings"
	"time"
)

func callAgentStream(ctx context.Context, prompt string, enableSearch bool) {
	flushChannel(ttsManagerChan)
	llmStart := time.Now()

	// 策略：联网搜索用 Max(准确但慢)，普通闲聊用 Turbo(极快)
	modelName := llmModelFast
	if enableSearch {
		modelName = llmModelSearch
		log.Println("LLM: 检测到时效性需求，已动态开启联网搜索...")
	}

	systemPrompt := "你是智能助手。仅在用户【明确要求播放音乐】（如“放首歌”、“听周杰伦”）时，才在回复末尾添加 [PLAY: 歌名]（随机播放用 [PLAY: RANDOM]）。" +
		"如果用户要求停止，加上 [STOP]。" +
		"回答天气、新闻、闲聊等普通问题时，【严禁】添加任何播放指令。"
	payload := map[string]interface{}{
		"model": modelName,
		"input": map[string]interface{}{
			"messages": []map[string]string{
				{"role": "system", "content": systemPrompt},
				{"role": "user", "content": prompt},
			},
		},
		"parameters": map[string]interface{}{
			"result_format":      "text",
			"incremental_output": true,
			"enable_search":      enableSearch, // 动态开关
		},
	}

	jsonBody, _ := json.Marshal(payload)
	req, _ := http.NewRequestWithContext(ctx, "POST", llmURL, bytes.NewReader(jsonBody))
	req.Header.Set("Authorization", "Bearer "+dashAPIKey)
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("X-DashScope-SSE", "enable")

	resp, err := insecureClient.Do(req)
	if err != nil {
		log.Printf("❌ LLM: 请求失败: %v", err)
		musicMgr.Unduck()
		return
	}
	defer resp.Body.Close()

	scanner := bufio.NewScanner(resp.Body)
	var fullTextBuilder strings.Builder
	var chunkBuffer strings.Builder
	var firstChunkSent = false

	fmt.Print("LLM 推理: ")

	for scanner.Scan() {
		select {
		case <-ctx.Done():
			return
		default:
		}
		line := scanner.Text()
		if !strings.HasPrefix(line, "data:") {
			continue
		}
		dataStr := strings.TrimPrefix(line, "data:")
		if strings.TrimSpace(dataStr) == "[DONE]" {
			break
		}

		var chunk struct {
			Output struct {
				Text string `json:"text"`
			} `json:"output"`
		}
		if err := json.Unmarshal([]byte(dataStr), &chunk); err == nil && chunk.Output.Text != "" {
			clean := cleanText(chunk.Output.Text)
			if clean == "" {
				continue
			}
			fmt.Print(clean)
			fullTextBuilder.WriteString(clean)
			chunkBuffer.WriteString(clean)

			// 动态调整首包断句阈值：联网搜索时降低阈值以减少用户焦虑
			threshold := 30
			if enableSearch {
				threshold = 15 // 搜索时只要有15个字或标点就立刻播报
			}

			if !firstChunkSent {
				if strings.ContainsAny(clean, "，。！？,.!?\n") || chunkBuffer.Len() > threshold {
					firstChunkSent = true
					sendChunk(&chunkBuffer)
				}
			} else {
				if strings.ContainsAny(clean, "，。！？,.!?\n") || chunkBuffer.Len() > 80 {
					sendChunk(&chunkBuffer)
				}
			}
		}
	}
	fmt.Println()
	log.Printf("⏱LLM 推理结束，总耗时: %v", time.Since(llmStart))

	// 处理剩余文本
	sendChunk(&chunkBuffer)
	ttsManagerChan <- "[[END]]"

	// 指令解析逻辑
	fullText := fullTextBuilder.String()
	if strings.Contains(fullText, "[STOP]") {
		musicMgr.Stop()
	}
	if matches := regexp.MustCompile(`(?i)\[PLAY:\s*(.*?)\]`).FindStringSubmatch(fullText); len(matches) > 1 {
		musicMgr.SearchAndPlay(strings.TrimSpace(matches[1]))
	}
}

// 辅助函数：发送文本块到 TTS
func sendChunk(buf *strings.Builder) {
	text := regexp.MustCompile(`\[.*?\]`).ReplaceAllString(buf.String(), "")
	if strings.TrimSpace(text) != "" {
		ttsManagerChan <- strings.TrimSpace(text)
	}
	buf.Reset()
}

File: py_ai_box/llm.py
import json
import re
import ssl
import time
import urllib.request

from . import state
from . import config_runtime
from .intent import clean_text


def call_agent_stream(cancel_event, prompt: str, enable_search: bool) -> None:
    from .util import flush_queue
    flush_queue(state.tts_manager_queue)
    llm_start = time.time()

    model_name = config_runtime.llm_model_fast
    if enable_search:
        model_name = config_runtime.llm_model_search
        print("LLM: 检测到时效性需求，已动态开启联网搜索...")

    system_prompt = (
        "你是智能助手。仅在用户【明确要求播放音乐】（如“放首歌”、“听周杰伦”）时，才在回复末尾添加 [PLAY: 歌名]（随机播放用 [PLAY: RANDOM]）。"
        "如果用户要求停止，加上 [STOP]。"
        "回答天气、新闻、闲聊等普通问题时，【严禁】添加任何播放指令。"
    )
    payload = {
        "model": model_name,
        "input": {
            "messages": [
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": prompt},
            ],
        },
        "parameters": {
            "result_format": "text",
            "incremental_output": True,
            "enable_search": enable_search,
        },
    }

    data = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        config_runtime.llm_url,
        data=data,
        headers={
            "Authorization": f"Bearer {config_runtime.dash_api_key}",
            "Content-Type": "application/json",
            "X-DashScope-SSE": "enable",
        },
    )
    ctx = ssl._create_unverified_context()
    try:
        resp = urllib.request.urlopen(req, context=ctx, timeout=60)
    except Exception as err:
        print(f"❌ LLM: 请求失败: {err}")
        if state.music_mgr is not None:
            state.music_mgr.unduck()
        return

    full_text = []
    chunk_buffer = []
    first_chunk_sent = False
    print("LLM 推理: ", end="", flush=True)

    for raw in resp:
        if cancel_event.is_set():
            resp.close()
            return
        line = raw.decode("utf-8", errors="ignore").strip()
        if not line.startswith("data:"):
            continue
        data_str = line[len("data:"):].strip()
        if data_str == "[DONE]":
            break
        try:
            chunk = json.loads(data_str)
        except Exception:
            continue
        text = chunk.get("output", {}).get("text", "")
        if not text:
            continue
        clean = clean_text(text)
        if not clean:
            continue
        print(clean, end="", flush=True)
        full_text.append(clean)
        chunk_buffer.append(clean)

        threshold = 15 if enable_search else 30
        buf_text = "".join(chunk_buffer)
        if not first_chunk_sent:
            if re.search(r"[，。！？,.!?\n]", clean) or len(buf_text) > threshold:
                first_chunk_sent = True
                send_chunk(buf_text)
                chunk_buffer = []
        else:
            if re.search(r"[，。！？,.!?\n]", clean) or len(buf_text) > 80:
                send_chunk(buf_text)
                chunk_buffer = []

    print()
    print(f"⏱LLM 推理结束，总耗时: {time.time() - llm_start:.3f}s")

    if chunk_buffer:
        send_chunk("".join(chunk_buffer))
    state.tts_manager_queue.put("[[END]]")

    full_text_str = "".join(full_text)
    if "[STOP]" in full_text_str and state.music_mgr is not None:
        state.music_mgr.stop()
    m = re.search(r"(?i)\[PLAY:\s*(.*?)\]", full_text_str)
    if m and state.music_mgr is not None:
        play_target = m.group(1).strip()
        print(f"[MUSIC-DBG] play_token={play_target}")
        state.music_mgr.search_and_play(play_target)
    resp.close()


def send_chunk(text: str) -> None:
    text = re.sub(r"\[.*?\]", "", text).strip()
    if text:
        state.tts_manager_queue.put(text)
